### PR TITLE
refactor!: use Socket.IO client instead of websockets for run-time data

### PR DESCRIPTION
**Pull request by eeberhard**
## Description

- https://github.com/aica-technology/api/issues/94

This PR solves the issue by replacing the websocket client with Socket.IO client. Breaking change, should be released as version 2.0.0 once the backend API is released as v3.0.0.

## Review guidelines

Estimated Time of Review: x minutes

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

**Review by LouisBrunner — APPROVED** [review]

**Review by JasperTan97 — APPROVED** [review]

File: python/src/aica_api/sio_client.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                    if elapsed >= timeout:
+                        raise TimeoutError
+                    received = sio.receive(timeout=timeout - elapsed)
+            except TimeoutError:
+                break
+            else:
+                if event == '*' or event == received[0]:
+                    try:
+                        if callback(data):
+                            return data
```
Comment (eeberhard) [inline python/src/aica_api/sio_client.py:62]:
```suggestion
                        if callback(received[1]):
                            return received[1]
```

**Review by jjenscodee — APPROVED** [review]

File: python/pyproject.toml [hunk 2]
```diff
@@ -13,7 +13,7 @@ readme = "README.md"
 requires-python = ">=3.7"
 dependencies = [
     "requests >= 2.28.1",
-    "websocket-client >= 1.4.2"
+    "python-socketio >= 5.11.0"
```
Comment (eeberhard) [inline python/pyproject.toml:16]: From @JasperTan97's testing:
```suggestion
    "python-socketio[client] >= 5.11.0"
```

Comment (JasperTan97) [reply python/pyproject.toml:16]: Don't forget to commit the suggestion!

Comment (eeberhard) [reply python/pyproject.toml:16]: yes indeed. is it fine with you if I don't merge this PR until we actually release v3?

**Comment by domire8** [comment]
Can this move forward as well now?

File: python/src/aica_api/client.py [hunk 3]
```diff
@@ -40,30 +41,37 @@ def _endpoint(self, endpoint=''):
...
-        if not new_version:
-            print(f'The detected API version v{api_version} is older than the minimum API version v2.0.0 supported by '
-                  f'this client')
-        return new_version
+
+        if api_version.startswith('3'):
+            return True
+        elif api_version.startswith('2'):
+            print(f'The detected API version v{api_version} is older than the minimum API version v3.0.0 supported by '
+                  f'this client (v{self.client_version()}). Please install Python API client version v1.2.0 or later'
```
Comment (domire8) [inline python/src/aica_api/client.py:69]: Is `later` the correct word here? I would understand it as > 1.2.0

**Review by domire8 — APPROVED** [review]
For me this still looks good, I will test again in the next couple of days but it can be merged before since the release will happen later

File: python/src/aica_api/client.py (hunk 3 above)
Comment (eeberhard) [reply python/src/aica_api/client.py:69]: Good point, since we know we aren't release anything between 1.2 and 2.0 we can make this more specific

Comment (eeberhard) [reply python/src/aica_api/client.py:69]:
```suggestion
                  f'this client (v{self.client_version()}). Please install Python API client version v1.2.0 '
```

**Comment by JasperTan97** [comment]
I've been using this since Jan 9 or 10 and had no issues with the API

**Pull request merged**